### DISCUSSION TITLE
MolToSmarts() command removes unspecified bonds from the molecule

### DISCUSSION THREAD

**Discussion by OmeirK** [rdkit/rdkit#7832] [General] [answered]
Hello,

I am trying to write some code where I read a molecule from a mol file and then automatically create a SMARTS pattern which only preserves a specific list of "anchor" atoms. 

At the moment, my code uses a list of indices for which I want to preserve, and an RDKit molecule object as the input. 

As far as I can tell, the code does the following:
1) Convert all non-anchor atoms to a dummy atom (with arbitrary atomic number of 0)
2) Check bonds between all atoms in the molecule. If a non-aromatic bond connects non-anchor atoms, edit the bond type to be a generic "Unspecified" bond type. 
3) Write a SMARTS pattern with the anchor atoms preserved, and non-anchor atoms and bonds converted to a more generic format. 

It appears that the atom editing part of the code is working as intended. However upon modifying bonds, the "Unspecified" bonds are not denoted with a "~", as ins standard for SMARTS patterns. This is an issue to me, as I want to use this SMARTS pattern to query a large database, and the types of bonds which connect atoms is an important detail for the search. 

My code and output are pasted below:
```
    anchor_ats = [0, 2, 7]
    mol = Chem.MolFromSmiles('[H]N([H])c1nc2c(ccn2[H])c(=O)[nH]1')
    mol_h = Chem.AddHs(mol, onlyOnAtoms=anchor_ats)
    print(Chem.MolToSmiles(mol_h))

    rwmol = Chem.RWMol(mol_h)
    
    # Step 1: Replace non-anchor atoms with a generic atom [Chem.Atom(0)]
    #print(anchor_ats)
    n_ats = rwmol.GetNumAtoms()
    for idx in range(0, n_ats):
        at = rwmol.GetAtomWithIdx(idx)
        elem = at.GetSymbol()

        if idx not in anchor_ats:
            if elem  != 'H':
                rwmol.ReplaceAtom(idx, Chem.Atom(0))

    # Convert the edited molecule to a smarts pattern to check it
    smarts = Chem.MolToSmarts(rwmol)

    print(f'SMARTS after atom editing: {smarts}')

    # Step 2: Replace bonds to all aromatic rings with a generic UNSPECIFIED bond
    for idx in range(0, n_ats):
        at = rwmol.GetAtomWithIdx(idx)
        if idx in anchor_ats:
            continue
        
        neighbors = at.GetNeighbors()
        
        # Replace non-aromatic bonds to non-anchor atoms with an "unspecified" bond
        for n in neighbors:
            n_idx = n.GetIdx()
            
            if n_idx in anchor_ats:
                continue
            
            bond = rwmol.GetBondBetweenAtoms(idx, n_idx)
            bond_type = bond.GetBondType()
            print(idx, n_idx, bond, bond_type, bond.GetIsAromatic(), at.GetSymbol(), n.GetSymbol())
            if bond.GetIsAromatic() == False:
                print('^ Replace this bond')
                bond.SetBondType(Chem.BondType.UNSPECIFIED)

    # View the smarts pattern with edited bonds
    smarts = Chem.MolToSmarts(rwmol)
    print(f'SMARTS after bond editing: {smarts}')
```

Output:

> [H]N([H])c1nc2c(ccn2[H])c(=O)[nH]1
`SMARTS after atom editing: [#7](-[#0]1:[#7]:[#0]2:[#0](:[#0]:[#0]:[#7]:2-[H]):[#0](=[#0]):[#0]:1)(-[H])-[H]`
> 1 10 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74e40> AROMATIC True * *
> 3 4 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74dd0> AROMATIC True * *
> 4 3 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74f90> AROMATIC True * *
> 4 5 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74f20> AROMATIC True * *
> 4 8 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74dd0> AROMATIC True * *
> 5 4 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74d60> AROMATIC True * *
> 5 6 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74c10> AROMATIC True * *
> 6 5 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74f20> AROMATIC True * *
> 8 4 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74d60> AROMATIC True * *
> 8 9 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74f90> DOUBLE False * *
> ^ Replace this bond
> 8 10 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74f20> AROMATIC True * *
> 9 8 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74e40> UNSPECIFIED False * *
> ^ Replace this bond
> 10 8 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74dd0> AROMATIC True * *
> 10 1 <rdkit.Chem.rdchem.Bond object at 0x7f6917e74c10> AROMATIC True * *
` SMARTS after bond editing: [#7](-[#0]1:[#7]:[#0]2:[#0](:[#0]:[#0]:[#7]:2-[H]):[#0]([#0]):[#0]:1)(-[H])-[H]`

Based on the output, it appears as though the correct bond was identified, and changed to an UNSPECIFIED bond type. However, when I convert the modified molecule to a SMARTS pattern, it appears as though (=[#0]) has been converted to ([#0]). From what I can tell, this is treated as a single bond when doing substructure matching. Is there any additional step I can take to ensure that it is written as (~[#0]) in the final SMARTS pattern?

Any help would be appreciated!

**Comment by ptosco** [comment, marked as answer]
@OmeirK You need to make it a `BOND_NULL_QUERY` bond, which in Python you may do as follows:

```
# You only need to define this once
BOND_NULL_QUERY_BOND = Chem.MolFromSmarts("*~*").GetBondWithIdx(0)
```
And change
```
bond.SetBondType(Chem.BondType.UNSPECIFIED)
```
into
```
rwmol.ReplaceBond(bond.GetIdx(), BOND_NULL_QUERY_BOND)
```
You will then get
```
SMARTS after bond editing: [#7](-[#0]1:[#7]:[#0]2:[#0](:[#0]:[#0]:[#7]:2-[H]):[#0](~[#0]):[#0]:1)(-[H])-[H]
```

**Reply by OmeirK** [reply]
Thanks so much! This was very helpful
